### PR TITLE
Add reduced support to reduceBy and reduceRight

### PR DESCRIPTION
**Pull request by morgsmccauley**
#2662

### CODE REVIEW

**Comment by CrossEye** [comment]
I'm in favor of this.  Anyone else, @ramda/core?

**Comment by GingerPlusPlus** [comment]
Bench:

```
~/JS/ramda (master *+$=)
$ npm run bench

> ramda@0.26.1 bench D:\Wojciech\JS\ramda
> node scripts/benchRunner

┌────────────────────────┬────────────────────────┬─────────────────┐
│ reduceRight            │ Hertz                  │ Margin of Error │
├────────────────────────┼────────────────────────┼─────────────────┤
│ reduceRight(add, 0, n… │ 16,477,354             │ 0.46%           │
├────────────────────────┼────────────────────────┼─────────────────┤
│ reduceRight(add, 0)(n… │ 2,541,277              │ 0.76%           │
├────────────────────────┼────────────────────────┼─────────────────┤
│ reduceRightAdd(nums)   │ 5,562,776              │ 0.98%           │
├────────────────────────┼────────────────────────┼─────────────────┤
│ native reduceRight     │ 1,132,187              │ 0.26%           │
└────────────────────────┴────────────────────────┴─────────────────┘

~/JS/ramda (pr/morgsmccauley/2663 *$=)
$ npm run bench

> ramda@0.25.0 bench D:\Wojciech\JS\ramda
> node scripts/benchRunner

┌────────────────────────┬────────────────────────┬─────────────────┐
│ reduceRight            │ Hertz                  │ Margin of Error │
├────────────────────────┼────────────────────────┼─────────────────┤
│ reduceRight(add, 0, n… │ 15,501,691             │ 0.47%           │
├────────────────────────┼────────────────────────┼─────────────────┤
│ reduceRight(add, 0)(n… │ 2,448,696              │ 0.74%           │
├────────────────────────┼────────────────────────┼─────────────────┤
│ reduceRightAdd(nums)   │ 5,402,445              │ 1.11%           │
├────────────────────────┼────────────────────────┼─────────────────┤
│ native reduceRight     │ 1,148,233              │ 0.22%           │
└────────────────────────┴────────────────────────┴─────────────────┘
```

**Comment by morgsmccauley** [comment]
Happy to do some more work to get this merged, alternatively I can close if it's no longer required :)

**Comment by CrossEye** [comment]
@morgsmccauley:

Sorry, like a few other issues, this one got dropped at some point.  If you want to try to fix the merge issues, this can definitely be included.

**Comment by morgsmccauley** [comment]
@CrossEye I'll take a look at the conflicts 👍

**Comment by adispring** [comment]
@morgsmccauley Do you have time to resolve the conflicts? So that ramda can merge this pr.
@CrossEye

**Comment by morgsmccauley** [comment]
@adispring yes, sorry for the delay, I'll try get this resolved shortly

**Comment by morgsmccauley** [comment]
@adispring i've resolved the conflicts. Let me know if anything else is required before merging 😄

File: source/reduceBy.js [hunk 1]
```diff
@@ -11,6 +12,9 @@ import _xreduceBy from './internal/_xreduceBy.js';
  * the String-returning function `keyFn` on each element and reduces the elements
  * of each group to a single value via the reducer function `valueFn`.
  *
+ * The value function receives two values: *(acc, value)*. It may use
+ * [`R.reduced`](#reduced) to shortcut the iteration.
```
Comment (adispring) [inline source/reduceBy.js:16]: `shortcut` -> `short circuit` ?

File: source/reduceRight.js [hunk 2]
```diff
@@ -10,7 +10,8 @@ import _curry3 from './internal/_curry3.js';
  * right to the left.
  *
  * The iterator function receives two values: *(value, acc)*, while the arguments'
- * order of `reduce`'s iterator function is *(acc, value)*.
+ * order of `reduce`'s iterator function is *(acc, value)*. `reduceRight` may use [`reduced`](#reduced)
+ * to shortcut the iteration.
```
Comment (adispring) [inline source/reduceRight.js:14]: as above, `shortcut` -> `short circuit`

Comment (morgsmccauley) [reply source/reduceRight.js]: done

File: source/reduceBy.js (hunk 1 above)
Comment (morgsmccauley) [reply source/reduceBy.js]: done

**Review by adispring — APPROVED** [review]

**Comment by adispring** [comment]
@morgsmccauley Thank you for your work, this looks good to me. 
@CrossEye Anything else?

**Comment by CrossEye** [comment]
:herb:

This looks great to me.  Thanks Morgan!

**Comment by morgsmccauley** [comment]
❤️

**Pull request merged**